### PR TITLE
[Validator] Add allowSpaces and allowLowerCase option in IBAN/BIC constraints

### PR DESCRIPTION
**Pull request by alamirault**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no  
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #28232
| License       | MIT
|Doc PR     | [https://github.com/symfony/symfony-docs/pull/11000](https://github.com/symfony/symfony-docs/pull/11000)

This PR allow to add options for BIC and IBAN constraints. I added allowSpaces and allowLowerCase options.
By default BIC allow spaces and disallow lower case, IBAN allow spaces and allow lower case.
It not same in order to avoid BC.

### CODE REVIEW

File: src/Symfony/Component/Validator/Tests/Constraints/IbanValidatorTest.php [hunk 1]
```diff
@@ -51,6 +51,7 @@ public function getValidIbans()
         return [
             ['CH9300762011623852957'], // Switzerland without spaces
             ['CH93  0076 2011 6238 5295 7'], // Switzerland with multiple spaces
+            ['ch93  0076 2011 6238 5295 7'], // Switzerland in lower case
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/Tests/Constraints/IbanValidatorTest.php]: Please add a testcase only with lower case and without spaces

**Comment by OskarStark** [comment]
You could validate a BIC and IBAN against each other, is it still possible when using different options on BIC and IBAN?

File: src/Symfony/Component/Validator/Constraints/BicValidator.php [hunk 2]
```diff
@@ -72,6 +72,15 @@ public function validate($value, Constraint $constraint)
             throw new UnexpectedValueException($value, 'string');
         }
 
+        if (!$constraint->allowSpaces && preg_match('/\s/', $value)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/BicValidator.php]: Using `\s` contradicts with replacing only spaces just below. The check should be `false !== strpos($value, ' ')`

**Comment by alamirault** [comment]
Tests certainly failed due to merge in twig repo, maybe [https://github.com/twigphp/Twig/pull/2848](https://github.com/twigphp/Twig/pull/2848) ?

**Comment by javiereguiluz** [comment]
I've been checking the case-sensitivity details for BIC and IBAN, which is surprisingly hard to find:

1) IBAN: 2 uppercase letters for the country + 2 digits + up to 30 alphanumeric characters for the bank account number. The last 30-character part is **case-insensitive**: https://en.wikipedia.org/wiki/International_Bank_Account_Number
2) BIC: it's not clear. The standard doesn't mention it: https://www.swift.com/standards/data-standards/bic#BICvalidationUserGuide but all validator you can find in other languages assume that it's **case-sensitive** and only uppercase letters are accepted. This (big) PDF contains the full list of all valid BIC codes in the world. All of them are uppercase: https://www.iso9362.org/documents/ISOBIC.pdf

So, I don't think we should make the case-sensitive behavior an option. This is a BIC + IBAN validator ... we cannot change those standards to validate what we want.

**Comment by ro0NL** [comment]
So perhaps it's better to leverage #26484 in userland..

**Comment by nicolas-grekas** [comment]
Always the same story: are validators also normalizers? If one wants to validate that it has only already normalized values in their db, then the check would be case sensitive and it should fail on whitespaces. Now, if one wants to accept user input and be robust (following the robustness principle), then the validator should accept both spaces and lowercase letters, and normalize them to uppercase+spaceless.

Both approaches are valid. Which domain is covered by this class?

**Comment by fabpot** [comment]
A validator should only validate the official way. So, adding options to deviate from that is a no go to me.

**Comment by nicolas-grekas** [comment]
Thank you @alamirault 
I'm going to close as explained.
On my side, I think we should make a clear distinction when the validator is in "strict" vs "filtering" mode.
That should be a concept that applies to almost all validators. (strict means the input is exactly the same as the output, after normalization)

**Pull request closed without merging**